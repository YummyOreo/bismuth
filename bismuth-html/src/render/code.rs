use syntect::easy::HighlightLines;
use syntect::highlighting::{Color, ThemeSet};
use syntect::html::highlighted_html_for_string;
use syntect::parsing::SyntaxSet;
use syntect::Error;

fn init() -> (SyntaxSet, ThemeSet) {
    let ps = SyntaxSet::load_defaults_newlines();
    let ts = ThemeSet::load_defaults();

    (ps, ts)
}

pub fn highlight(lang: String, code: String) -> Result<String, Error> {
    let (ps, ts) = init();
    let syntax = ps.find_syntax_by_extension(&lang).unwrap();

    highlighted_html_for_string(&code, &ps, syntax, &ts.themes["InspiredGitHub"])
}

#[cfg(test)]
mod test {

    use super::*;
    macro_rules! snapshot {
        ($content:tt) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../../testdata/output/render/code/");
            settings.bind(|| {
                insta::assert_snapshot!($content);
            });
        };
    }

    #[test]
    fn test() {
        let code = "fn test() {\n\tlet a = 10;\n}".to_string();
        let html = highlight("rs".to_string(), code).unwrap();
        snapshot!(html);
    }

    #[test]
    fn test_inline() {
        let code = "let test = 10".to_string();
        let html = highlight("rs".to_string(), code).unwrap();
        snapshot!(html);
    }
}
