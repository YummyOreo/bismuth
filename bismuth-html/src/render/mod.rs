#![allow(dead_code, unused)]
use bismuth_parser::{
    error,
    tree::{Ast, Element, Kind},
    Parser,
};
use std::path::PathBuf;

mod element;
mod code;

pub trait Render {
    fn render<T: Render + Clone>(&mut self, content: &[T]) -> String;
}

#[derive(Clone)]
pub struct Renderer {
    pub parser: Parser,

    output: String,
    path: PathBuf,

    current_elements: Vec<element::HtmlElement>,
}

impl Renderer {
    pub fn new(parser: Parser) -> Self {
        let path = PathBuf::from(
            parser
                .metadata
                .frontmatter
                .get_path()
                .expect("Should have a path"),
        );
        Self {
            parser,
            output: String::new(),
            path,
            current_elements: vec![],
        }
    }
}

