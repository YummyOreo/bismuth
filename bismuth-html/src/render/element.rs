use crate::render::Render;

#[derive(Clone)]
pub enum HtmlElement {
    Paragraph,
    Text { text: String },
    Bold,
    Italic,
    Blockquote,

    Link { text: String, link: String },
    Filelink { text: String, link: String },

    ListItem { level: i32 },
    NumList { level: i32, num: String },

    InlineCode { text: String },
    Blockcode { text: String, lang: String },

    InlineLaTeX { tex: String },
    BlockLaTeX { tex: String },

    Header { level: i8 },

    HorizontalRule,
    LineBreak,

    // todo: Later impl this, after above
    CustomElm,
}

impl Render for HtmlElement {
    fn render<T: Render + Clone>(&mut self, content: &[T]) -> String {
        let mut inside = String::new();
        for (index, r) in content.iter().enumerate() {
            inside.push_str(&r.clone().render(content.split_at(index + 1).1));
        }
        let (start, end) = match self {
            Self::Paragraph => ("<p>".to_owned(), "</p>".to_owned()),
            Self::Text { text } => (text.clone(), "".to_owned()),
            Self::Bold => ("<b>".to_string(), "</b>".to_string()),
            Self::Italic => ("<i>".to_string(), "</i>".to_string()),
            Self::Blockquote => ("<blockquote>".to_string(), "</blockquote>".to_string()),
            Self::Link { text, link } => (
                format!("<a href=\"{}\" target=\"_blank\">{}", link, text),
                "</a>".to_string(),
            ),
            Self::Filelink { text, link } => (
                format!("<img src=\"{}\" alt=\"{}\">", link, text),
                "".to_string(),
            ),

            Self::ListItem { level } => (
                format!("<li class=\"level-{}\">", level),
                "</li>".to_string(),
            ),
            Self::NumList { level, num } => (
                format!("<li class=\"level-{}\">{}.", level, num),
                "</li>".to_string(),
            ),

            // TODO: InlineCode + BLockcode + InlineLaTeX + BlockLaTeX
            Self::HorizontalRule => ("<hr>".to_string(), "".to_string()),
            Self::LineBreak => ("<br>".to_string(), "".to_string()),
            _ => ("".to_owned(), "".to_owned()),
        };
        todo!();
    }
}
