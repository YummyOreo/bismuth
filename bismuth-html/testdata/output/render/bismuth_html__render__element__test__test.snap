---
source: bismuth-html/src/render/element.rs
expression: rendered
---
<p>this is a test<a href="example.com" target="_blank">test</a></p>
